x-pack/legacy/plugins/spaces/server/lib/copy_to_spaces/copy_to_spaces.ts

```diff
@@ -0,0 +1,101 @@
+ */
+
+import { SavedObjectsClientContract, SavedObjectsService } from 'src/core/server';
+import { Readable } from 'stream';
+import { SavedObjectsClientProviderOptions } from 'src/core/server';
+import { SpacesClient } from '../spaces_client';
+import { Rereadable } from './lib/rereadable_stream';
+import { spaceIdToNamespace } from '../utils/namespace';
+import { CopyOptions } from './types';
+import { canImportIntoSpace } from './lib/can_import_into_space';
+import { CopyResponse } from './types';
+import { getEligibleTypes } from './lib/get_eligible_types';
+import { createEmptyFailureResponse } from './lib/create_empty_failure_response';
+
+export const COPY_TO_SPACES_SAVED_OBJECTS_CLIENT_OPTS: SavedObjectsClientProviderOptions = {
+  excludedWrappers: ['spaces'],
+};
+
+export function copySavedObjectsToSpacesFactory(
+  spacesClient: SpacesClient,
+  savedObjectsClient: SavedObjectsClientContract,
+  savedObjectsService: SavedObjectsService
+) {
+  const { importExport, types, schema } = savedObjectsService;
+  const eligibleTypes = getEligibleTypes({ types, schema });
+
+  const exportRequestedObjects = async (
+    sourceSpaceId: string,
+    options: Pick<CopyOptions, 'includeReferences' | 'objects'>
+  ) => {
+    const objectStream = await importExport.getSortedObjectsForExport({
+      namespace: spaceIdToNamespace(sourceSpaceId),
+      includeReferencesDeep: options.includeReferences,
+      objects: options.objects,
+      savedObjectsClient,
+      types: eligibleTypes,
+      exportSizeLimit: importExport.objectLimit,
+    });
+    return objectStream;
+  };
+
+  const importObjectsToSpace = async (
+    spaceId: string,
+    objectsStream: Readable,
+    options: CopyOptions
+  ) => {
+    try {
+      const { canImport, reason } = await canImportIntoSpace(
+        spaceId,
+        spacesClient,
+        savedObjectsClient
+      );
+
+      if (!canImport) {
+        return createEmptyFailureResponse([{ error: { type: reason!, spaceId } }]);
+      }
+
+      const importResponse = await importExport.importSavedObjects({
+        namespace: spaceIdToNamespace(spaceId),
+        objectLimit: importExport.objectLimit,
+        overwrite: options.overwrite,
+        savedObjectsClient,
+        supportedTypes: eligibleTypes,
+        readStream: objectsStream,
+      });
+
+      return {
+        success: importResponse.success,
+        successCount: importResponse.successCount,
+        errors: importResponse.errors,
+      };
+    } catch (error) {
+      return createEmptyFailureResponse([error]);
+    }
+  };
+
+  const copySavedObjectsToSpaces = async (
+    sourceSpaceId: string,
+    destinationSpaceIds: string[],
+    options: CopyOptions
+  ): Promise<CopyResponse> => {
+    const response: CopyResponse = {};
+
+    const objectsStream = await exportRequestedObjects(sourceSpaceId, options);
+    const rereadableStream = objectsStream.pipe(new Rereadable());
```

Comment:
Let's chat about this import/export stream stuff tomorrow.

Reply:
Discussed on Zoom -- we don't like the fact that the import/export APIs are stream based, but don't actually allow for streaming the data in an efficient manner. Essentially, the entire import and export payloads are kept in memory. Rather than addressing it as part of this initiative, we would rather make it more explicit that we are holding the entire export payload in memory while the copy-to-space operation is in progress.

To do this, we will remove the `Rereadable` stream introduced in this PR, and instead read the stream to completion before feeding the contents to the `import` operation.

Implemented in 79cc29fc005